### PR TITLE
DistinctOperation now returns a batch Cursor

### PR DESCRIPTION
**Pull request by rozza**
JAVA-1624

### CODE REVIEW

File: driver-async/src/main/com/mongodb/async/client/MongoCollectionImpl.java [hunk 1]
```diff
@@ -180,37 +176,17 @@ public void count(final Object filter, final CountOptions options, final SingleR
...
-                        }
-                        callback.onResult(distinctList, null);
-                    } catch (Throwable tr) {
-                        callback.onResult(null, new MongoException("Error when decoding distinct results", tr));
-                    }
-                }
-            }
-        }));
+
+        return new OperationIterable<Object>(operation, readPreference, executor);
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoCollectionImpl.java]: Seems like this is a behavior change.  The previous implementation used the Document codec to transform each BsonValue into it's corresponding value type defined by the DocumentCodec.  Now I think it will return BsonValue.

File: driver-async/src/test/functional/com/mongodb/async/client/SmokeTestSpecification.groovy [hunk 2]
```diff
@@ -67,6 +67,16 @@ class SmokeTestSpecification extends FunctionalSpecification {
...
         then: 'The count is zero'
         run(collection.&count) == 0
+
+        then: 'InsertMany documents'
+        run(collection.&insertMany, [new Document('id', 'a'), new Document('id', 1),
+                                     new Document('id', new Document('b', 'c')),
+                                     new Document('id', new Document('list', [2, 'd', new Document('e', 3)]))]) == null
+
+        then: 'Distinct'
+        run(collection.distinct('id', new Document()).&into, []) == ['a', 1, new Document('b', 'c'),
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/SmokeTestSpecification.groovy]: This would seem to indicate that I'm wrong, but I'm not clear how it works

File: driver-core/src/main/com/mongodb/operation/DistinctOperation.java [hunk 3]
```diff
@@ -18,30 +18,50 @@
...
+ *
  * <p>When possible, the distinct command uses an index to find documents and return values.</p>
  *
  * @mongodb.driver.manual reference/command/distinct Distinct Command
  * @since 3.0
  */
-public class DistinctOperation implements AsyncReadOperation<BsonArray>, ReadOperation<BsonArray> {
+public class DistinctOperation implements AsyncReadOperation<AsyncBatchCursor<Object>>, ReadOperation<BatchCursor<Object>> {
+    private static final String VALUES = "values";
+    private static final Decoder<Document> DECODER = new DocumentCodec();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/DistinctOperation.java]: OK, now I see what you did.  Any reason not to pass the Decoder into the constructor?  This would be the first time we reference the Document class in any operation.

File: driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java [hunk 4]
```diff
@@ -58,7 +58,7 @@
         this.namespace = firstQueryResult.getNamespace();
         this.limit = limit;
         this.batchSize = batchSize;
-        this.decoder = notNull("decoder", decoder);
+        this.decoder = decoder;
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java]: Why is this allowed to be null now?

File: driver/src/main/com/mongodb/DBCollection.java [hunk 5]
```diff
@@ -1067,17 +1066,8 @@ public List distinct(final String fieldName, final DBObject query) {
...
-
-        List distinctList = new ArrayList();
-        for (BsonValue value : distinctArray) {
-            BsonDocument document = new BsonDocument("value", value);
-            DBObject obj = getDefaultDBObjectCodec().decode(new BsonDocumentReader(document), DecoderContext.builder().build());
-            distinctList.add(obj.get("value"));
-        }
-
-        return distinctList;
+        return new OperationIterable<Object>(new DistinctOperation(getNamespace(), fieldName).filter(wrap(query)),
```
Comment (jyemin) [inline driver/src/main/com/mongodb/DBCollection.java]: This is the problem with hard-coding DocumentCodec in DistinctOperation, as this method can now return Document instances instead of DBObject instances.

File: driver/src/main/com/mongodb/client/MongoCollection.java [hunk 6]
```diff
@@ -163,7 +163,7 @@
      * @return a non-null list of distinct values
      * @mongodb.driver.manual reference/command/distinct/ Distinct
      */
-    List<Object> distinct(String fieldName, Object filter, DistinctOptions distinctOptions);
+    MongoIterable<Object> distinct(String fieldName, Object filter, DistinctOptions distinctOptions);
```
Comment (jyemin) [inline driver/src/main/com/mongodb/client/MongoCollection.java]: I'm thinking now we'll need an overload that takes a Class<T> (as with aggregate, etc), so users can control which codec is used to decode each value.

**Comment by jyemin** [comment]
First pass complete.  Theme is that DistinctOperation needs to have a configurable Codec or else return a Cursor<BsonValue>

File: driver-async/src/main/com/mongodb/async/client/MongoCollectionImpl.java (hunk 1 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MongoCollectionImpl.java]: Same behaviour as internally it uses the document codec - which needs to be configurable.

File: driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java (hunk 4 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java]: Because we dont need a decoder and we dont have a `decoder<Object> `

Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java]: Because there is no get-more for distinct.  Got it.  It's a bit of a wart though.

Comment (rozza) [inline driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java]: Worth a new class? `ListResultBatchCursor`

Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java]: Since it's not a public class I'm not really worried enough to do that.

File: driver/src/main/com/mongodb/client/MongoCollection.java (hunk 6 above)
Comment (rozza) [inline driver/src/main/com/mongodb/client/MongoCollection.java]: I cant see how this can be accomplished at the moment as we have to embed the values in a BsonDocument and then decode.  With a `decoder<T>` we lose the ability to get those values out of the decoded item :(

**Comment by rozza** [comment]
I removed the codec from `DistinctOperation` and now we map in the collection to the correct type.

Without being able to decode a BsonArray directly I can't see how we can use a provided codec.

File: driver-async/src/test/unit/com/mongodb/async/client/MongoCollectionSpecification.groovy [hunk 7]
```diff
@@ -211,13 +211,16 @@ class MongoCollectionSpecification extends Specification {
...
+        }
+        def executor = new TestOperationExecutor([asyncCursor, asyncCursor])
         def collection = new MongoCollectionImpl(namespace, Document, codecRegistry, readPreference, writeConcern, executor)
+        def filter = new BsonDocument()
         def futureResultCallback = new FutureResultCallback<List<Object>>()
 
         when:
-        collection.distinct('test', filter, futureResultCallback)
+        collection.distinct('test', filter).into([], futureResultCallback)
         futureResultCallback.get()
```
Comment (jyemin) [inline driver-async/src/test/unit/com/mongodb/async/client/MongoCollectionSpecification.groovy]: Would this test be better if it asserted something about the contents of the collection returned by into?

File: driver-core/src/test/functional/com/mongodb/operation/DistinctOperationSpecification.groovy [hunk 8]
```diff
@@ -47,10 +47,10 @@ class DistinctOperationSpecification extends OperationFunctionalSpecification {
 
         when:
         DistinctOperation op = new DistinctOperation(getNamespace(), 'name')
-        def result = op.execute(getBinding());
+        def result = op.execute(getBinding()).next();
 
         then:
-        result.toList().sort() == new BsonArray([new BsonString('Pete'), new BsonString('Sam')])
+        result == [new BsonString('Pete'), new BsonString('Sam')]
```
Comment (jyemin) [inline driver-core/src/test/functional/com/mongodb/operation/DistinctOperationSpecification.groovy]: I'm confused here.  Shouldn't next() return a single BsonString?

File: driver/src/main/com/mongodb/DBCollection.java [hunk 9]
```diff
@@ -286,9 +285,9 @@ public WriteResult insert(final List<? extends DBObject> documents, final WriteC
     /**
      * <p>Insert documents into a collection. If the collection does not exists on the server, then it will be created. If the new document
      * does not contain an '_id' field, it will be added.</p>
-     * 
-     * <p>If the value of the continueOnError property of the given {@code InsertOptions} is true, 
-     * that value will override the value of the continueOnError property of the given {@code WriteConcern}. Otherwise, 
+     *
```
Comment (jyemin) [inline driver/src/main/com/mongodb/DBCollection.java]: Can you revert the unnecessary formatting changes to comments?

**Comment by jyemin** [comment]
Something is wrong, I think. It appears that DistinctOperation is now returning a BatchCursor containing a single BsonArray instance, instead of a series of BsonValue instances, one for each element of the array.

File: driver-async/src/test/unit/com/mongodb/async/client/MongoCollectionSpecification.groovy (hunk 7 above)
Comment (rozza) [inline driver-async/src/test/unit/com/mongodb/async/client/MongoCollectionSpecification.groovy]: I dont think so - because we have these test cases are covered elsewhere, all we care about here is that the DistinctOperation is correctly configured.  OperationIterable and DistinctOperation are tested separately and we have a functional test doing end to end testing.

File: driver-core/src/test/functional/com/mongodb/operation/DistinctOperationSpecification.groovy (hunk 8 above)
Comment (rozza) [inline driver-core/src/test/functional/com/mongodb/operation/DistinctOperationSpecification.groovy]: No its a batchCursor

File: driver/src/main/com/mongodb/DBCollection.java (hunk 9 above)
Comment (rozza) [inline driver/src/main/com/mongodb/DBCollection.java]: Its difficult as I have remove whitespace on save - which should be the default, at least has been ever since I got the updated intellij configs.

I could turn it off and then manually redo this whole pull request - but I'm not sure its worth it? - as we shouldn't have trailing whitespace.

Comment (jyemin) [inline driver/src/main/com/mongodb/DBCollection.java]: ok, i give.

File: driver-core/src/test/functional/com/mongodb/operation/DistinctOperationSpecification.groovy (hunk 8 above)
Comment (jyemin) [inline driver-core/src/test/functional/com/mongodb/operation/DistinctOperationSpecification.groovy]: Right, but each call to next() should return one BsonString.

Comment (rozza) [inline driver-core/src/test/functional/com/mongodb/operation/DistinctOperationSpecification.groovy]: Not according the the batchCursor interface - next returns List<T>

Comment (rozza) [inline driver-core/src/test/functional/com/mongodb/operation/DistinctOperationSpecification.groovy]: See: https://github.com/rozza/mongo-java-driver/blob/3.0.x/driver-core/src/main/com/mongodb/operation/BatchCursor.java#L56

**Comment by rozza** [comment]
Closing in favour of #58 a typesafe variant
